### PR TITLE
allow the filter function to be async

### PR DESCRIPTION
**Pull request by NeoPhi**
@helfer This was the changed I mentioned to allow the filter function to be async.

### CODE REVIEW

File: src/test/tests.ts [hunk 1]
```diff
@@ -234,6 +241,31 @@ describe('SubscriptionManager', function() {
     });
   });
 
+  it('can use an async filter function', function(done) {
+    const query = `subscription Filter2($filterBoolean: Boolean){
+       testFilter(filterBoolean: $filterBoolean)
+      }`;
+    const callback = function(err, payload){
```
Comment (DxCx) [inline src/test/tests.ts]: you are ignoring `err`, is this intended?

File: src/test/tests.ts [hunk 2]
```diff
@@ -234,6 +241,31 @@ describe('SubscriptionManager', function() {
     });
   });
 
+  it('can use an async filter function', function(done) {
```
Comment (DxCx) [inline src/test/tests.ts]: i might be missing something, how do you test async here?

Comment (NeoPhi) [inline src/test/tests.ts]: The publish below is for `Filter2` which [as defined above](https://github.com/apollographql/graphql-subscriptions/blob/d4dfbedca6fec24ddd6794d212adab5cf5ac9977/src/test/tests.ts#L125) uses `setTimeout` to delay returning the result.

File: src/test/tests.ts (hunk 1 above)
Comment (NeoPhi) [inline src/test/tests.ts]: As the test case is structured an err would be unexpected but easy enough to check.

File: src/test/tests.ts (hunk 2 above)
Comment (DxCx) [inline src/test/tests.ts]: yeah i've seen it, but it means the test is actually making sure promises work, 
and not that things are running in parallel..
name is abit confusing.. 🗡

Comment (NeoPhi) [inline src/test/tests.ts]: I don't see it as that. If you revert the changes to `pubsub.ts` this test will fail since and async filter is always truthy even if it ultimately returns a false value. If you have a suggestion about how to restructure the test please let me know.

Comment (DxCx) [inline src/test/tests.ts]: oh no, i don't mind at all that this is what you are testing.
i would just rename the test abit to something like "can use promise inside filter function"
or something like that.. since usually when testing 'async', the testing is about being run in parallel

File: src/pubsub.ts [hunk 3]
```diff
@@ -182,26 +182,29 @@ export class SubscriptionManager {
...
-                if (typeof options.context === 'function') {
-                    contextPromise = new Promise((resolve) => {
-                        resolve(options.context());
-                    });
-                } else {
-                    contextPromise = Promise.resolve(options.context);
-                }
-                contextPromise.then((context) => {
-                    if (!filter(rootValue, context)) {
+                Promise.resolve().then(() => {
```
Comment (Urigo) [inline src/pubsub.ts]: I think you can optimize this code and do a better use of Promise chaining - this way you can avoid the inner promises and the empty `resolve()`. you can also merge the two `catch` that does the same thing.

Comment (NeoPhi) [inline src/pubsub.ts]: I wasn't able to see an easy way to do that since the context needs to be in the closure for two unrelated calls.

Comment (Urigo) [inline src/pubsub.ts]: I see, but you think that something like this won't do the trick?
```ts
const onMessage = (rootValue) => {
    const contextResult = typeof options.context === 'function' ? options.context() : context;

    return Promise
        .resolve(contextResult)
        .then(context => filter(rootValue, context))
        .then(filterResult => {
            if (!filterResult) return;

            return execute(
                this.schema,
                parsedQuery,
                rootValue,
                context,
                options.variables,
                options.operationName
            );
        })
        .then(executionResult => {
            options.callback(null, executionResult);
        })
        .catch((error) => {
            options.callback(error);
        });
}
```

Comment (NeoPhi) [inline src/pubsub.ts]: `context` is not defined in the call to execute.

Comment (Urigo) [inline src/pubsub.ts]: @NeoPhi , so we can resolve it any way, and do the same for context, and then use `Promise.all`... something like that: 

```ts
const onMessage = (rootValue) => {
    const contextResult = typeof options.context === 'function' ? options.context() : context;

    return Promise
        .resolve(contextResult)
        .then(context => Promise.all([ Promise.resolve(filter(rootValue, context)), Promise.resolve(context) ]) )
        .then(promisesResults=> {
            const filterResult = promisesResults[0];
            const context = promisesResults[1];

            if (!filterResult) return;

            return execute(
                this.schema,
                parsedQuery,
                rootValue,
                context,
                options.variables,
                options.operationName
            );
        })
        .then(executionResult => {
            options.callback(null, executionResult);
        })
        .catch((error) => {
            options.callback(error);
        });
}
```

File: src/test/tests.ts (hunk 2 above)
Comment (NeoPhi) [inline src/test/tests.ts]: Updated the test name.

File: src/pubsub.ts (hunk 3 above)
Comment (NeoPhi) [inline src/pubsub.ts]: Woke up with the same thought. Although you can't unnest the `executionResult` otherwise the callback will be called with `(null, undefined)` unless that is desired.

**Comment by Urigo** [comment]
thank you @NeoPhi !
